### PR TITLE
Feat.metadata.debug

### PR DESCRIPTION
**Pull request by aneksteind**
Adds debug info representing original line of MIR to PDG snapshots. This will aid in debugging incorrect PDG node types and relations by avoiding the need to manually cross reference.

### CODE REVIEW

File: pdg/src/snapshots/c2rust_pdg__tests__analysis_test_pdg_snapshot.snap [hunk 1]
```diff
@@ -3,999 +3,999 @@ source: pdg/src/main.rs
...
+	n[7]:  addr.load  n[6]  => _   @ bb4[13]: fn simple; _11 = ((*_1).0: i32)                       
+	n[8]:  field.0    n[1]  => _   @ bb4[14]: fn simple; ((*_5).0: i32) = move _11                  
+	n[9]:  addr.store n[8]  => _   @ bb4[14]: fn simple; ((*_5).0: i32) = move _11                  
+	n[10]: field.1    n[3]  => _   @ bb4[16]: fn simple; ((*_1).1: u64) = const 9_u64               
+	n[11]: addr.store n[10] => _   @ bb4[16]: fn simple; ((*_1).1: u64) = const 9_u64               
+	n[12]: field.0    n[3]  => _12 @ bb4[18]: fn simple; _12 = ((*_1).0: i32)                       
+	n[13]: addr.load  n[12] => _   @ bb4[18]: fn simple; _12 = ((*_1).0: i32)                       
+	n[14]: field.1    n[3]  => _13 @ bb4[21]: fn simple; _13 = &raw const ((*_1).1: u64)            
+	n[15]: addr.load  n[14] => _   @ bb4[21]: fn simple; _13 = &raw const ((*_1).1: u64)            
+	n[16]: addr.load  n[3]  => _   @ bb4[24]: fn simple; _14 = &raw const (*_1)
```
Comment (spernsteiner) [inline pdg/src/snapshots/c2rust_pdg__tests__analysis_test_pdg_snapshot.snap:112]: Assuming the new statement listing is accurate, these two `LoadAddr`s should not be here, since `x = &(*y).1` does not access memory.

File: pdg/src/snapshots/c2rust_pdg__tests__analysis_test_pdg_snapshot.snap [hunk 2]
```diff
@@ -3,999 +3,999 @@ source: pdg/src/main.rs
...
-	n[28]: field.0    n[27] => _95  @ bb51[2]:  fn exercise_allocator;
-	n[29]: addr.load  n[28] => _    @ bb51[2]:  fn exercise_allocator;
-	n[30]: copy       n[1]  => _97  @ bb49[15]: fn exercise_allocator;
-	n[31]: offset[3]  n[30] => _96  @ bb49[20]: fn exercise_allocator;
-	n[32]: field.0    n[31] => _95  @ bb51[2]:  fn exercise_allocator;
-	n[33]: addr.load  n[32] => _    @ bb51[2]:  fn exercise_allocator;
-	n[34]: copy       n[1]  => _106 @ bb59[7]:  fn exercise_allocator;
-	n[35]: copy       n[34] => _105 @ bb59[8]:  fn exercise_allocator;
-	n[36]: free       n[35] => _104 @ bb59[9]:  fn exercise_allocator;
+	n[0]: &_61 _    => _60 @ bb29[7]:  fn exercise_allocator; _60 = &raw const (*_61)
```
Comment (spernsteiner) [inline pdg/src/snapshots/c2rust_pdg__tests__analysis_test_pdg_snapshot.snap:240]: Does `&_61` represent `AddrOfLocal(_61)`?  If so, this node is also incorrect - it should be a `Copy` from `_61` instead.

File: pdg/src/snapshots/c2rust_pdg__tests__analysis_test_pdg_snapshot.snap [hunk 3]
```diff
@@ -3,999 +3,999 @@ source: pdg/src/main.rs
...
+	n[0]: &_17 _    => _16 @ bb3[11]: fn invalid; _16 = &raw const (*_17)                              
+	n[1]: copy n[0] => _15 @ bb3[12]: fn invalid; _15 = move _16 as *const u8 (Pointer(ArrayToPointer))
+	n[2]: copy n[1] => _14 @ bb3[14]: fn invalid; _14 = move _15 as *const i8 (Misc)                   
+	n[3]: copy n[2] => _1  @ bb0[0]:  fn printf; _13 = pointers::printf(move _14, move _18)            
 }
 nodes_that_need_write = []
 
 g {
-	n[0]: field.0 _ => _18 @ bb3[18]: fn invalid;
+	n[0]: field.0 _ => _18 @ bb3[18]: fn invalid; _18 = ((*(*_19)).0: i32)
```
Comment (spernsteiner) [inline pdg/src/snapshots/c2rust_pdg__tests__analysis_test_pdg_snapshot.snap:425]: I think the input to this node should be a `LoadValue`.  This statement loads a pointer value `p` from `*_19`, then computes `q = &(*p).0`, and finally loads from `*q`.  So there should be two chains, `_19 -> LoadAddr` and `LoadValue -> Field -> LoadAddr`.

Also, the node having destination `_18` is incorrect.  Given the RHS, the type of `_18` must be `i32` (i.e. not a pointer type), so the PDG should not track anything about it.

File: pdg/src/snapshots/c2rust_pdg__tests__analysis_test_pdg_snapshot.snap [hunk 4]
```diff
@@ -3,999 +3,999 @@ source: pdg/src/main.rs
...
-	n[12]: addr.store  n[11] => _       @ bb3[4]:  fn connection_accepted;
-	n[13]: copy        n[8]  => _0      @ bb3[7]:  fn connection_accepted;
-	n[14]: copy        n[13] => _13     @ bb5[0]:  fn lighttpd_test;      
-	n[15]: copy        n[14] => _17     @ bb5[7]:  fn lighttpd_test;      
-	n[16]: copy        n[15] => _2      @ bb0[0]:  fn connection_close;   
-	n[17]: field.1     n[16] => _5      @ bb0[4]:  fn connection_close;   
-	n[18]: addr.load   n[17] => _       @ bb0[4]:  fn connection_close;   
-	n[19]: field.0     n[16] => _8      @ bb1[7]:  fn connection_close;   
-	n[20]: addr.load   n[19] => _       @ bb1[7]:  fn connection_close;   
+	n[0]:  alloc       _     => _5      @ bb1[2]:  fn connection_accepted; _5 = pointers::malloc(move _6)
```
Comment (spernsteiner) [inline pdg/src/snapshots/c2rust_pdg__tests__analysis_test_pdg_snapshot.snap:560]: The output seems to produce a lot of unnecessary whitespace at end of line in some cases

**Review by spernsteiner — APPROVED** [review]
I only looked at the snapshot, not the code, but this seems like a great change.  It makes it much easier to spot mismatches between the PDG and the code.

File: pdg/src/snapshots/c2rust_pdg__tests__analysis_test_pdg_snapshot.snap (hunk 2 above)
Comment (aneksteind) [reply pdg/src/snapshots/c2rust_pdg__tests__analysis_test_pdg_snapshot.snap:240]: that's fixed in #533

**Comment by aneksteind** [comment]
@spernsteiner thanks for the review. There are no functional changes here so I will create separate issues for your feedback. I'm glad this change is already proving useful!

File: pdg/src/snapshots/c2rust_pdg__tests__analysis_test_pdg_snapshot.snap (hunk 4 above)
Comment (aneksteind) [reply pdg/src/snapshots/c2rust_pdg__tests__analysis_test_pdg_snapshot.snap]: thanks, this was a bug

File: pdg/src/graph.rs [hunk 5]
```diff
@@ -155,7 +158,7 @@ impl Node {
         let fn_ = function;
         write!(
             f,
-            "{kind}{sep}{src}{sep}=>{sep}{dest}{sep}@{sep}{bb_stmt}:{sep}fn {fn_};"
+            "{kind}{sep}{src}{sep}=>{sep}{dest}{sep}@{sep}{bb_stmt}:{sep}fn {fn_};{sep}{debug}"
```
Comment (kkysen) [inline pdg/src/graph.rs:161]:
```suggestion
            "{kind}{sep}{src}{sep}=>{sep}{dest}{sep}@{sep}{bb_stmt}:{sep}fn {fn_};{sep}{debug};"
```

End with a `;` to make syntax highlighting better.

File: pdg/src/snapshots/c2rust_pdg__tests__analysis_test_pdg_snapshot.snap [hunk 6]
```diff
@@ -3,999 +3,999 @@ source: pdg/src/main.rs
...
-	n[0]: &_1 _ => _29 @ bb15[11]: fn main;
+	n[0]: &_1 _ => _29 @ bb15[11]: fn main; _29 = &mut _1
 }
 nodes_that_need_write = []
 
 g {
-	n[0]: copy _    => _34 @ bb16[5]: fn main;  
-	n[1]: copy n[0] => _2  @ bb0[0]:  fn expect;
+	n[0]: copy _    => _34 @ bb16[5]: fn main;   _34 = &(*_35)                                                                                 
+	n[1]: copy n[0] => _2  @ bb0[0]:  fn expect; _31 = std::result::Result::<std::ffi::CString, std::ffi::NulError>::expect(move _32, move _34)
```
Comment (kkysen) [inline pdg/src/snapshots/c2rust_pdg__tests__analysis_test_pdg_snapshot.snap:42]: Can we prune paths like `rustc` error messages do, especially for types in the prelude?  I'm not sure how that's done, but maybe there's an exposed function in the `rustc` crates.

File: pdg/src/snapshots/c2rust_pdg__tests__analysis_test_pdg_snapshot.snap (hunk 1 above)
Comment (aneksteind) [reply pdg/src/snapshots/c2rust_pdg__tests__analysis_test_pdg_snapshot.snap]: addressed in #565

File: pdg/src/graph.rs [hunk 7]
```diff
@@ -119,6 +119,8 @@ pub struct Node {
     pub kind: NodeKind,
     /// The `Node` that produced the input to this operation.
     pub source: Option<NodeId>,
+
+    pub debug: String,
```
Comment (kkysen) [inline pdg/src/graph.rs:123]: Maybe call this `debug_info` instead.

File: analysis/runtime/src/mir_loc.rs [hunk 8]
```diff
@@ -169,6 +169,8 @@ pub struct EventMetadata {
     pub destination: Option<MirPlace>,
     /// Destination func [`DefPathHash`] of [`Event`](crate::events::Event).
     pub transfer_kind: TransferKind,
+    /// Any string useful for debugging
+    pub debug: String
```
Comment (kkysen) [inline analysis/runtime/src/mir_loc.rs:173]: Maybe call this `debug_info` instead.

File: analysis/runtime/src/mir_loc.rs [hunk 9]
```diff
@@ -169,6 +169,8 @@ pub struct EventMetadata {
     pub destination: Option<MirPlace>,
     /// Destination func [`DefPathHash`] of [`Event`](crate::events::Event).
     pub transfer_kind: TransferKind,
+    /// Any string useful for debugging
```
Comment (kkysen) [inline analysis/runtime/src/mir_loc.rs:172]:
```suggestion
    /// Any string useful for debugging.
```

File: dynamic_instrumentation/src/instrument.rs [hunk 10]
```diff
@@ -203,14 +205,19 @@ impl<'tcx> Visitor<'tcx> for InstrumentationAdder<'_, 'tcx> {
...
+                    .debug_mir(location)
                     .add_to(self);
 
                 if is_region_or_unsafe_ptr(value_ty) {
-                    location.statement_index += 1;
-                    self.loc(location, store_value_fn)
+                    let instrumentation_location = Location {
+                        statement_index: location.statement_index + 1,
+                        ..location
+                    };
```
Comment (kkysen) [inline dynamic_instrumentation/src/instrument.rs:215]: Can we add a `Location::inc` method?

File: dynamic_instrumentation/src/point/build.rs [hunk 11]
```diff
@@ -139,6 +139,36 @@ impl<'tcx> InstrumentationBuilder<'_, 'tcx> {
...
+                            s.push_str(&format!("{:?}", arg));
+                        }
+                        s.push(')');
+                        s
+                    }
+                    _ => String::from(""),
+                }
+            } else {
+                format!("{:?}", block.statements[loc.statement_index])
+            }
```
Comment (kkysen) [inline dynamic_instrumentation/src/point/build.rs:167]: I don't like the unnecessary `destination.unwrap()`, though I get why it;s convenient here with the `match` `if`.  Maybe there's a better way to refactor this.

File: analysis/runtime/src/mir_loc.rs (hunk 9 above)
Comment (kkysen) [reply analysis/runtime/src/mir_loc.rs:172]: Fixed in https://github.com/immunant/c2rust/pull/566/commits/0e2477d482a5b9f58a6b904f66243d6070d1af24.

File: analysis/runtime/src/mir_loc.rs (hunk 8 above)
Comment (kkysen) [reply analysis/runtime/src/mir_loc.rs:173]: Fixed in https://github.com/immunant/c2rust/pull/566/commits/a9fd15ff6a20f7e05758ed8aa83743b9ba6bfb77.

File: analysis/runtime/src/mir_loc.rs [hunk 12]
```diff
@@ -177,6 +179,7 @@ impl Default for EventMetadata {
             source: None,
             destination: None,
             transfer_kind: TransferKind::None,
```
[the remark this answers is not in the document]
Comment (kkysen) [reply analysis/runtime/src/mir_loc.rs:181]: Fixed in https://github.com/immunant/c2rust/pull/566/commits/db9726d84b31db6cfb1ad0afd30aeb13e49aef23.

File: pdg/src/graph.rs (hunk 7 above)
Comment (kkysen) [reply pdg/src/graph.rs:123]: Fixed in https://github.com/immunant/c2rust/pull/566/commits/a9fd15ff6a20f7e05758ed8aa83743b9ba6bfb77.

File: pdg/src/graph.rs (hunk 5 above)
Comment (kkysen) [reply pdg/src/graph.rs:161]: Fixed in https://github.com/immunant/c2rust/pull/566/commits/7e3105321218231f871a8110609f922335763f79.

File: dynamic_instrumentation/src/instrument.rs (hunk 10 above)
Comment (kkysen) [reply dynamic_instrumentation/src/instrument.rs:215]: Fixed in https://github.com/immunant/c2rust/pull/566/commits/d1429553d6eff8d2b89968c06666e06577344cd3 (used the existing `Location::successor_within_block`, which does exactly the same thing).

File: dynamic_instrumentation/src/point/build.rs (hunk 11 above)
Comment (kkysen) [reply dynamic_instrumentation/src/point/build.rs:167]: Fixed in https://github.com/immunant/c2rust/pull/566/commits/cc477cbb94a1fb09a4d8f3d1f3412b3ae35f3727.
